### ISSUE TITLE
createDataSpacenet.py crashed when transforming data to PASCAL VOL2012

### ISSUE THREAD

**Issue by yghlc** [SpaceNetChallenge/utilities#22]
I used createDataSpacenet.py to transform SpanceNet to PASCAL VOL 2012, the script crashed with the following message:
` Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)`
The codes position of crashing is in "labelTools.py": 
```
        featureDefn = innerBufferLayer.GetLayerDefn()
        bufferDist = srcRaster.GetGeoTransform()[1]*bufferSizePix
        for idx, feature in enumerate(source_layer):
            ingeom = feature.GetGeometryRef()  #cashed at second loop
            geomBufferOut = ingeom.Buffer(bufferDist) 
            geomBufferIn  = ingeom.Buffer(-bufferDist)
            print(geomBufferIn.ExportToWkt())
            print(geomBufferIn.IsEmpty())
            print(geomBufferIn.IsSimple())
```

It worked fine in the first loop but crashed in the second, the images I used is 
 AOI_2_Vegas_Train/RGB-PanSharpen/RGB-PanSharpen_AOI_2_Vegas_img1.tif'

I also tried to use other images, some of them work well, but some of them failed.

The parameters I used are: aws_SpaceNet/un_gz/AOI_2_Vegas_Train --srcImageryDirectory RGB-PanSharpen --outputDirectory aws_SpaceNet/voc_format/AOI_2_Vegas_Train/ --annotationType PASCALVOC2012 --imgSizePix 400 --convertTo8Bit --spacenetVersion 2
The gdal I used is version 2.1.3, I don't know whether this matter. I upgraded to this version because of the requirement of other packages.

I don't know what's going on, many thanks for any suggestion.

**Comment by yghlc** [comment]
After more test, I found this issue could due to the quality of geojson provided in SpaceNet. 
I test on two geojson files: buildings_AOI_2_Vegas_img1.geojson and buildings_AOI_2_Vegas_img101.geojson.

For buildings_AOI_2_Vegas_img1.geojson, there is a point in this file, so the file contains two layers, one is polygon layer, the other is point layer. After I deleted the point in this file, the codes in labelTools.py can process this file successfully.

For buildings_AOI_2_Vegas_img101.geojson, it only has one layer, that is polygon layer, but there is a very small triangle in the polygon layer, then the codes in labelTools.py failed in processing this file. After I deleted the small triangle, the codes can handle this file.

I already spent one day on this issue, and I think there are more geojson files can cause this issue. And I don't have time to edit geojson files one by one.  The codes is not robust enough or the quality of spaceNet is not good enough.

Any suggestion?

**Comment by dlindenbaum** [comment]
Hello,

Generally we've found that newer versions of GDAL do not have this problem.  Try upgrading your gdal binaries and gdal python

I believe this is the same error as seen in #16 
_________

For posterity some questions about your hardware/software setup
What OS?
What GDAL version is python using?
in python execute
```
from osgeo import gdal
gdal.VersionInfo()
```
What gdal executables do you have
on command line?
```
gdalinfo --version

GDAL 1.11.3, released 2015/09/16
```

Thank you,
Dave

**Comment by yghlc** [comment]
@dlindenbaum Thank you for your reply.
The OS is:
    Ubuntu 16.04
    Python 2.7.12
GDAL version:
```
>>> from osgeo import gdal
>>> gdal.VersionInfo()
'2010300'
```
```
gdalinfo --version
GDAL 2.1.3, released 2017/20/01
```
Maybe I downgrade the gdal to 1.11.3 can solve this problem.

Actually, I already removed all the files can cause createDataSpacenet.py crashed. I wrote script to do this. In the end, it's almost 10% of the files can cause createDataSpacenet.py crashed.

**Comment by dlindenbaum** [comment]
I think it's that your python bindings for GDAL are old. 2010300, is an older version.   Try updating those.  I don't think you need to downgrade the GDAL binaries.    


```
apt-get install python-gdal
```

Thank you,

**Comment by dlindenbaum** [comment]
Hello, 

are you using Anaconda/conda?

**Comment by dlindenbaum** [comment]
Hello,

It appears that one problem may be with that Certain versions of GDAL are taking the field "Id" in the GeoJSON and using that as a File ID.  since that Id field is not unique in the dataset it is causing errors.  You can remove all attributes from a geoJSON.  I believe that will allow you to run using all the data.  I'll work on getting an updated list of geoJSONs that have the unused attributes deleted.  

Below is a command that will hopefully fix the geojsons for this use case.

``` 
mkdir tmp
basename -s.geojson *.geojson | xargs -n1 -I % ogr2ogr -f GeoJSON tmp/%tmp.geojson %.geojson -select "NAME"
```

**Comment by yghlc** [comment]
@dlindenbaum Thanks for your work.
I didn't use Anaconda/conda. I use the default python on Ubuntu 16.04.

When I tried to upgrade the python bindings for GDAL, it said it is the newest version. See the following message.   I think "2010300"  isn't old version, it means 2.1.3, not in the year of 2010.
Also, on my MacBookPro, the "gdal.VersionInfo()" in python output '2010100', and "gdalinfo --version" output "GDAL 2.1.1, released 2016/07/07". 
```
hlc@Cryo06:~$ sudo apt-get install python-gdal
Reading package lists... Done
Building dependency tree
Reading state information... Done
python-gdal is already the newest version (2.1.3+dfsg-1~xenial2).
0 upgraded, 0 newly installed, 0 to remove and 10 not upgraded.
```

**Comment by dlindenbaum** [comment]
@yghlc 
You are correct.  I made a mistake on interpreting GDAL Versions.   I did some exploration.  Something occurred with the newer GDAL bindings which is causing the error.  I've been able to reproduce it.  

I have not figured out a way to fix it in code but I will work to fix the data and have it re-uploaded.

In the mean time I believe that the below will solve the problem, 

1.  Remove all Fields from geojson except NAME and geometry
```
cd /path/to/geoJsonDirectory/
mkdir tmp
basename -s.geojson *.geojson | xargs -n1 -I % ogr2ogr -f GeoJSON tmp/%.geojson %.geojson -select "partialBuilding, partialDec"

```
2. Remove old geojsons from folder
```
rm *.geojson
```
3.  move new geojsons back into the directory 
```
mv tmp/*.geojson .
```

I'll update this threat when he code works or data has been updated.

Thank you for the issue notice.

Dave

**Comment by dlindenbaum** [comment]
@yghlc I believe I fixed the error and it should run as expected After the last merge.

**Comment by yghlc** [comment]
@dlindenbaum Many Thanks. I will run the codes and tell you my result this weekend because I have an exam tomorrow.

**Comment by yghlc** [comment]
@dlindenbaum 
Hi, Dave
   I ran your updated codes, it works for all the data. Many thanks!
Lingcao

**Comment by yghlc** [comment]
fix #22
